File: cpp/snippets/binary_trees.cpp
#include "../common_header.h"
#include<iostream>
#include<vector>
#include "../utils/utils.h"
#include<algorithm>

/*
          1
    2            3
4       5   6           7
*/
TreeNode* create_binary_tree() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    return root;
}

TreeNode* tree(const std::vector<int>& vec, int start, int end) {
    int m = start + (end - start) / 2;
    TreeNode* node = new TreeNode(vec[m]);
    node->left = tree(vec, start, m-1);
    node->right = tree(vec, m+1, end);
    return node;
}

/*
create bst from an array
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
*/
TreeNode* array_to_binary_search_tree(const std::vector<int>& vec) {
    std::sort(vec.begin(), vec.end());
    TreeNode* node = tree(vec, 0, vec.size()-1);
    return node;
}

/*
https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/description/
*/
TreeNode* list_to_binary_search_tree(ListNode* root) {
    TreeNode* root = nullptr;
    return root;
}

#include<queue>
std::vector<int> bfs(TreeNode* root) {
    std::vector<int> result;
    if(!root) {
        return result;
    }
    std::queue<TreeNode*> q;
    q.push(root);
    while(!q.empty()) {
        auto node = q.front();
        result.push_back(node->val);
        q.pop();
        if(node->left) {
            q.push(node->left);
        }
        if(node->right) {
            q.push(node->right);
        }
    }
    return result;        
}

void inorder(TreeNode* root, std::vector<int>& results) {
    if(!root) {
        return;
    }
    inorder(root->left, results);
    results.push_back(root->value);
    inorder(root->right, results);
}

void preorder(TreeNode* root, std::vector<int>& results) {
    if(!root) {
        return;
    }
    results.push_back(root->value);
    preorder(root->left, results);
    preorder(root->right, results);    
}


void postorder(TreeNode* root, std::vector<int>& results) {
    if(!root) {
        return;
    }
    postorder(root->left, results);
    postorder(root->right, results);    
    results.push_back(root->value);
}

#include<stack>
/*  start with the current, dont fill the stack first.
    go to leftmost & keep putting in the stack.... 
    pop the left most.
    make current as the right of the left.
*/
void inorder_iterative(TreeNode* root, std::vector<int>& results) {
    if(!root) {
        return;
    }
    std::stack<TreeNode*> st;
    auto current = root;
    while(current || !st.empty()) {
        while(current) {
            st.push(current);
            current = current->left;
        }
        current = st.top();
        results.push_back(current->value);
        st.pop();
        current = current->right;
    }
}

/*  
root => left => right
put in stack...
iterate till stack is empty.
    pop from stack.
    put right,left 
*/
void preorder_iterative(TreeNode* root, std::vector<int>& results) {
    if(!root) {
        return;
    }
    std::stack<TreeNode*> st;
    st.push(root);
    while(!st.empty()) {
        auto node = st.top();
        results.push_back(node->value);
        st.pop();
        if(node->right) {
            st.push(node->right);
        }
        if(node->left) {
            st.push(node->left);
        }
    }
}

/*
    put the root in to stack
    while stack {
        root value
        left
        right
    }
    reverse
*/
void postorder_iterative(TreeNode* root, std::vector<int>& results) {
    std::stack<TreeNode*> st;
    st.push(root);
    while(!st.empty()) {
        auto node = st.top();
        results.push_back(node->value);
        st.pop();
        if(node->left) {
            st.push(node->left);
        }
        if(node->right) {
            st.push(node->right);
        }
    }
    std::reverse(results.begin(), results.end());
}

int height(TreeNode* root) {
    int h = 0;
    if(!root) {
        return 0;
    }
    h = 1 + std::max(height(root->left), height(root->right));
    return h;
}


bool isBalanced(TreeNode* root) {
        return checkHeight(root) != -1;
    }

int checkHeight(TreeNode* node) {
        if (!node) return 0; // Base case: empty tree has height 0
        
        // Check left subtree
        int leftHeight = checkHeight(node->left);
        if (leftHeight == -1) return -1; // Left subtree is unbalanced
        
        // Check right subtree
        int rightHeight = checkHeight(node->right);
        if (rightHeight == -1) return -1; // Right subtree is unbalanced
        
        // Check if current node is balanced
        if (std::abs(leftHeight - rightHeight) > 1) return -1;
        
        // Return height of current node
        return std::max(leftHeight, rightHeight) + 1;
    }


int find_max(TreeNode* root) {
    if(!root) {
        return INT_MIN;
    }
    int maximum = root->value;
    int l = find_max(root->left);
    int r = find_max(root->right);
    maximum = std::max(maximum, l);
    maximum = std::max(maximum, r);
    return maximum;
}


int main() {
    TreeNode* root = create_binary_tree();
    std::cout << "bfs" << std::endl;
    auto results = bfs(root);
    print_vector(results);
    //printBinaryTree(root);
    return 0;

    std::cout << "inorder" << std::endl;
    results.clear();
    inorder(root, results);
    print_vector(results);

    std::cout << "preorder" << std::endl;
    results.clear();
    preorder(root, results);
    print_vector(results);

    std::cout << "post-order" << std::endl;
    results.clear();
    postorder(root, results);
    print_vector(results);
    results.clear();

    std::cout << "height" << std::endl;
    int h = height(root);
    std::cout << h << std::endl;

    std::cout << "is_balanced" << std::endl;
    bool bal = is_balanced(root);
    std::cout << bal << std::endl;

    std::cout << "find_maximum" << std::endl;
    int max = find_max(root);
    std::cout << max << std::endl;

    //iterative traversal
    std::cout << "inorder_iterative" << std::endl;
    inorder_iterative(root, results);
    print_vector(results);  
    results.clear(); 

    std::cout << "preorder_iterative" << std::endl;
    preorder_iterative(root, results);
    print_vector(results);   
    results.clear(); 
    
    std::cout << "postorder_iterative" << std::endl;
    postorder_iterative(root, results);
    print_vector(results);   

    
}

File: python/snippets/binary_trees.py
#bfs on binary tree..
from tree_node import TreeNode
from tree_node import bfs
from collections import deque

#binary tree balanced => number of leaves = N/2 + 1

#bfs - in level traversal..
    #gives all the nodes at a particular level
#t: O(N)
#s: O(N)
def bfs_binary_tree(root: TreeNode):
    q = deque()
    q.append(root)
    while q:
        l = len(q)
        for _ in range(l):
            node = q.popleft()
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)


#dfs
#t: O(N)  => travel each node once
#s: O(H)    => recursion.






#binary search tree
'''
recursive
iterative
search: t: O(logN)
        s: O(H)
'''

#preorder

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Recursive Solution
def inorderTraversal_recursive(root):
    result = []
    
    def inorder(node):
        if not node:
            return
        inorder(node.left)    # Visit left subtree
        result.append(node.val)  # Visit current node
        inorder(node.right)   # Visit right subtree
    
    inorder(root)
    return result

# Iterative Solution
def inorderTraversal(root):
    result = []
    stack = []
    current = root
    
    while current or stack:
        # Reach the leftmost node
        while current:
            stack.append(current)
            current = current.left
        
        # Process current node
        current = stack.pop()
        result.append(current.val)
        
        # Move to right subtree
        current = current.right
    
    return result

# Example usage:
#     1
#      \
#       2
#      /
#     3
root = TreeNode(1)
root.right = TreeNode(2)
root.right.left = TreeNode(3)

print(inorderTraversal(root))  # Output: [1, 3, 2]



#post order


# Recursive Solution
def postorderTraversal_recursive(root):
    result = []
    
    def postorder(node):
        if not node:
            return
        postorder(node.left)    # Visit left subtree
        postorder(node.right)   # Visit right subtree
        result.append(node.val) # Visit current node
    
    postorder(root)
    return result

# Iterative Solution (using two stacks)
def postorderTraversal(root):
    if not root:
        return []
    
    result = []
    stack = [root]
    
    # First stack to process nodes, second to reverse order
    while stack:
        node = stack.pop()
        result.append(node.val)  # Add to result first (will be reversed)
        
        # Push left child first, then right
        # (so right will be processed before left in final result)
        if node.left:
            stack.append(node.left)
        if node.right:
            stack.append(node.right)
    
    return result[::-1]  # Reverse the result

# Example usage:
#     1
#      \
#       2
#      /
#     3
root = TreeNode(1)
root.right = TreeNode(2)
root.right.left = TreeNode(3)

print(postorderTraversal(root))  # Output: [3, 2, 1]





def preorder_traversal_recursive(root):
    """
    Recursive implementation of preorder traversal.

    Args:
        root: The root node of the binary tree.

    Returns:
        A list containing the preorder traversal of the tree.
    """
    result = []

    def preorder(node):
        if node:
            result.append(node.val)  # Visit the current node
            preorder(node.left)     # Traverse the left subtree
            preorder(node.right)    # Traverse the right subtree

    preorder(root)
    return result

def preorder_traversal_iterative(root):
    """
    Iterative implementation of preorder traversal.

    Args:
        root: The root node of the binary tree.

    Returns:
        A list containing the preorder traversal of the tree.
    """
    if not root:
        return []

    result = []
    stack = [root]

    while stack:
        node = stack.pop()
        result.append(node.val)

        # Push right child first, so left child is processed first.
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)

    return result

# Example usage:
# Constructing a sample binary tree
#        1
#       / \
#      2   3
#     / \
#    4   5

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

print("Recursive Preorder Traversal:", preorder_traversal_recursive(root))
print("Iterative Preorder Traversal:", preorder_traversal_iterative(root))